src/transformers/generation/logits_process.py

```diff
@@ -3342,3 +3342,108 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
+        scores = scores.reshape(-1, self.num_codebooks, self.codebook_size)
+        batch_size = scores.shape[0]
+
+        # we only look at the n-th last tokens to initialize the bos and eos delay patterns, where n is the delay pattern size
+        delay_pattern_size = len(self.delay_pattern)
+        input_ids = input_ids[:, -delay_pattern_size:]
+
+        # Initialize bos delay pattern
+        if self.bos_delay_pattern is None:
+            self.bos_delay_pattern = self.delay_pattern.repeat(batch_size, 1)
+            audio_bos_idxs = (input_ids == self.audio_bos_token_id).nonzero()
+
+            if len(audio_bos_idxs) > 0:
+                batch_idxs = audio_bos_idxs[:, 0]
+                is_first = torch.cat([batch_idxs.new_ones(1, dtype=torch.bool), batch_idxs[1:] != batch_idxs[:-1]])
+                min_bos_idxs = audio_bos_idxs[is_first]
+                current_after_bos = (delay_pattern_size - min_bos_idxs[:, 1]).unsqueeze(-1)
+                unique_batch_idxs = batch_idxs.unique().to(self.bos_delay_pattern.device)
+                self.bos_delay_pattern[unique_batch_idxs] = self.bos_delay_pattern[
+                    unique_batch_idxs
+                ] - current_after_bos.to(self.bos_delay_pattern.device)
+            else:
+                # there is no audio bos token,
+                self.bos_delay_pattern = torch.zeros_like(self.bos_delay_pattern)
+
+        # Initialize eos delay pattern
+        if self.eos_delay_pattern is None:
+            self.eos_delay_pattern = self.delay_pattern.repeat(batch_size, 1)
+            audio_eos_idxs = (input_ids == self.audio_eos_token_id).nonzero()
+
+            if len(audio_eos_idxs) > 0:
+                batch_idxs = audio_eos_idxs[:, 0]
+                is_first = torch.cat([batch_idxs.new_ones(1, dtype=torch.bool), batch_idxs[1:] != batch_idxs[:-1]])
+                min_eos_idxs = audio_eos_idxs[is_first]
+                current_after_eos = (delay_pattern_size - min_eos_idxs[:, 1]).unsqueeze(-1)
+                unique_batch_idxs = batch_idxs.unique().to(self.eos_delay_pattern.device)
+                self.eos_delay_pattern[unique_batch_idxs] = self.eos_delay_pattern[
+                    unique_batch_idxs
+                ] - current_after_eos.to(self.eos_delay_pattern.device)
+
+        # at each generation step, we decrement the bos delay pattern
+        row_mask = self.bos_delay_pattern >= 0
+        scores[(row_mask[..., None] & self.vocab_mask_bos).to(scores.device)] = -float("inf")
+        self.bos_delay_pattern[row_mask] -= 1
+
+        # when the audio eos token is generated, we decrement the eos delay pattern
+        self.eos_delay_pattern[input_ids[:, -1].to(self.eos_delay_pattern.device) == self.audio_eos_token_id] -= 1
+        row_mask = self.eos_delay_pattern <= 0
+        scores[(row_mask[..., None] & self.vocab_mask_eos).to(scores.device)] = -float("inf")
```

Comment:
don't know if float -inf won't have issue with bf16/fp16 for the rest of the scores?